prov/rxd: various updates/fixes

aingerson opened the pull request:
- Reposting fix
- Fix incorrect op
- Optimize retrying
- Optimize/fix unexpected messaging reposting
- Don't send acks if peer is uninitialized
- Get rid of pending_cnt

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -1011,8 +1011,11 @@ void rxd_ep_progress(struct util_ep *util_ep)
 		ret = rxd_ep_post_buf(ep);
+		if (ret)
+			break;
+	}
```
This change is fine, but please look at changing the code structure so that the variables that control the while loop are visibly updated inside the loop, rather than as a by-product hidden inside some other call.  Consider moving the while loop into the rxd_ep_post_buf call, for example.

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -1011,8 +1011,11 @@ void rxd_ep_progress(struct util_ep *util_ep)
-	while (ep->posted_bufs < ep->rx_size && !ret)
+	while (ep->posted_bufs < ep->rx_size) {
 		ret = rxd_ep_post_buf(ep);
+		if (ret)
```
Note that if there was some sort of more severe error that occurred, the progress function doesn't handle it (even to print out a message).

aingerson replied:
Are you talking about in general or specifically with posting the buffers? rxd_ep_post_buf prints a message if there was an error.

shefty replied:
I was talking general to this function.  ret is being overwritten here, so any severe error has been lost.  Note that I'm not requesting changes to this patch, but future possible enhancements.

shefty commented on prov/rxd/src/rxd_cq.c:
```diff
@@ -329,7 +329,10 @@ static int rxd_send_cts(struct rxd_ep *rxd_ep, struct rxd_rts_pkt *rts_pkt,
+	if (dlist_empty(&rxd_ep->peers[peer].unacked) ||
+	    rxd_get_base_hdr(container_of((&rxd_ep->peers[peer].unacked)->next,
+			     struct rxd_pkt_entry, d_entry))->type != RXD_RTS)
+		rxd_update_peer(rxd_ep, peer, rts_pkt->rts_addr);
```
The commit message for this change suggests a deeper problem.  I will check directly.

aingerson commented:
@shefty  Updated with alternate fix to the RTS-RTS problem. I don't think we can fully avoid not sending messages if the peer isn't active (dropped CTS on either side) so I think we need to just be ok with ignoring/not acking packets from uninitialized peers.
I replaced the weird update patch with one dealing with the data from an inactive peer and added another patch that fixes some scaling issues I was seeing. 
After this PR is merged, rxd should be able to scale up better (I've tested it on up to 64 ranks)

shefty commented:
Please see CI failure:
```
- name:   fi_getinfo_test -s n25-mlx5_0_1 n25-mlx5_0_1 -p "verbs;ofi_rxd"
  result: Fail
...
    Testing for fabric: IB-0xfe80000000000000, domain: mlx5_0-dgram, endpoint type: 3
    FAIL: source addrlen not set
    Running getinfo_no_hints4 [Test with node, service, FI_SOURCE flags and no hints]...[debug] fabtests:unit/getinfo_test.c:534: 
    Testing for fabric: verbs_any_fabric, domain: verbs_any_domain, endpoint type: 1
    FAIL: source addrlen not set
```

aingerson commented:
@shefty  I think our mlx nodes decided to go to lunch yesterday for some reason since there are a few failures that all happened at the same time. I couldn't reproduce the errors and I tested everything. It all seems ok.
